pymc3/model.py

```diff
@@ -800,7 +804,36 @@ def cont_vars(self):
         """All the continuous variables in the model"""
         return list(typefilter(self.vars, continuous_types))
 
-    def Var(self, name, dist, data=None, total_size=None):
+    def shape_from_dims(self, dims):
+        shape = []
+        if len(set(dims)) != len(dims):
+            raise ValueError('Can not contain the same dimension name twice.')
+        for dim in dims:
+            if dim not in self.coords:
+                raise ValueError('Unknown dimension name %s. All dimension '
+                                 'names must be specified in the `coords` '
+                                 'argument of the model or through a pm.Data '
+                                 'variable.' % dim)
+            shape.extend(self.coords[dim].shape)
+        return tuple(shape)
+
+    def add_coords(self, coords):
+        if coords is None:
+            return
+
+        for name in coords:
+            if not name.isidentifier():
+                raise ValueError('Invalid dimension name %s' % name)
```

Comment:
Is that actually true? I don't think that column names in pandas data frames need to be valid python identifiers, if you access them with strings.  I don't know what constraints `arviz` imposes.  I definitely work with scientists who have spaces in their column names.

Reply:
xarray does support spaces in var names, the netcdf standard discourages it though. We could be more permissive, but changing that back later would be hard. Personally, I quite dislike spaces in column names, so maybe I'm biased here though.
We might want to add some kind of separator to those names in the future (in many models the parameters and dimensions have some kind of a tree-like structure after all), and that would be much easier if we are restrictive now.

I don't think we would need a custom translation, you can always just rename the columns of the dataframe: `df.rename(columns=mapping)`.

Reply:
@aseyboldt OK. I still am not convinced that we should demand valid python identifiers. We don't demand this of our variable names, so why our dimensions and coordinates?  Certainly, if we *were* to demand this, we would need a **much better** error message in the `ValueError` that would make it clear that we are demanding Python identifiers here. Otherwise the poor user just knows that we don't like their name, but has no clue how to fix it.
The more I think about this, the less I like it.

Reply:
I already changed the message a bit.
To be honest I never liked that we accept strange variable names. :-)
But I see your point about consistency...

Reply:
I was concerned because I have been working with some biologists, and sometimes the column names I see definitely are not valid python identifiers.
Perhaps more concerning, some of these column names in CSV files are automatically generated by lab automation systems, not by humans at all.

I'm generally not in favor of adding yet another inconsistency to working with PyMC3, and I think if we do, we must make it absolutely obvious to the user what we want them to do to comply with our new restriction.

Reply:
I see what you mean.
I'll think a bit about it and also ask around on slack

Reply:
As we're reviving the PR, we should come to an agreement on this question.

My opinion is that we should enforce a good naming practice here. NetCDF discourages it and enforcing it here is an opportunity to "push the good practice up the stack".

My colleague just suggested to show a warning? (cc @AlexAndorra )

Reply:
I've changed my opinion - see the comment below.. We're updating the example notebook to showcase why..